src/osd/modules/input/input_evdev.cpp

```diff
@@ -0,0 +1,713 @@
+					device()->add_item(name, id, generic_button_get_state<uint8_t>, keystate+i);
+				}
+			}
+
+		if(fd < 0) {
+			char fname[16];
+			snprintf(fname, 16, "event%d", info.num);
+			fd = openat(dirfd, fname, O_RDONLY|O_NONBLOCK);
+			if(fd < 0) {
+				osd_printf_verbose("evdev: unable to open %s: %s\n", fname, strerror(errno));
+			}
+		}
+
+		mapi = 0;
+		for(int i=0; i<REL_MAX; i++)
+			if(info.relaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || relaxis_map[mapi].evdev_id) && relaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(relaxis_map[mapi].evdev_id == i) {
+					device()->add_item(relaxis_map[mapi].name, relaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, relaxis+i);
+				}
+			}
+
+		mapi = 0;
+		for(int i=0; i<ABS_MAX; i++)
+			if(info.absaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || absaxis_map[mapi].evdev_id) && absaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(absaxis_map[mapi].evdev_id == i) {
+					if(fd >= 0)
+						ioctl(fd, EVIOCGABS(i), &absaxis[i].absinfo);
+					device()->add_item(absaxis_map[mapi].name, absaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, &absaxis[i].normalized);
+				}
+			}
+	}
+
+	void poll(void) override
+	{
+		static input_event ie[256]; // try to consume as many as possible at once
+		while(fd >= 0) {
+			ssize_t rd = read(fd, ie, 256*sizeof(input_event));
+			if(rd < 0) {
+				if(errno==EWOULDBLOCK || errno==EAGAIN)
+					return;
+				// cope with errors here
+			}
+			if(rd < sizeof(input_event))
+				return;
+			int nie = rd/sizeof(input_event);
+			for(int i=0; i<nie; i++) {
+				switch(ie[i].type) {
+					case EV_KEY:
+						keystate[ie[i].code] = (ie[i].value>0)? 0x80: 0;
+						break;
+					case EV_REL:
+						relaxis[ie[i].code] += ie[i].value;
+						break;
+					case EV_ABS: {
+						  auto& axis = absaxis[ie[i].code];
+						  axis.normalized = normalize_absolute_axis(ie[i].value, axis.absinfo.minimum, axis.absinfo.maximum);
+						}
+						break;
+				}
+			}
+		}
+	}
+
+	void reset(void) override
+	{
+		memset(keystate, 0, KEY_MAX);
+		memset(relaxis, 0, REL_MAX*sizeof(int32_t));
+		memset(absaxis, 0, ABS_MAX*sizeof(absaxis[0]));
+	}
+};
+
+#define MAX_EVDEV_DEVICES 64
+
+class evdev_input
+{
+private:
+	bool initialized;
+
+public:
+	int  devinput;
+	evdev_device_info dev[MAX_EVDEV_DEVICES]; // 640k should be enough for anyone
+	int max_dev;
+
+public:
+	evdev_input(): initialized(false)
+	{
+	}
+
+	void init(void)
+	{
+		if(initialized)
+			return;
+		initialized = true;
+		osd_printf_verbose("evdev: enumerating devices\n");
+		devinput = open("/dev/input", O_RDONLY|O_CLOEXEC|O_DIRECTORY);
+		if(devinput < 0) {
+			return;
+		}
+
+		max_dev = 0;
+		for(int i=0; i<MAX_EVDEV_DEVICES; i++) {
+			dev[i].mapped = 0;
+			dev[i].type = EVDEV_NONE;
+		}
+
+		DIR* idir = fdopendir(devinput);
+		while(idir) {
+			dirent* de = readdir(idir);
+			if(!de)
+				break;
+			if(strncmp(de->d_name, "event", 5))
+				continue;
+			int dnum = atoi(de->d_name+5);
```

Comment:
`atoi` depends on the global locale – you should avoid it if possible.

Reply:
Is there a usable non-locale dependent function I can use instead?  I doubt I'm the only one finding myself gaving to decoding decimal from ascii.

Reply:
(Not an issue, the refactor eleminated all atoi() calls anyways)